client/components/date-range/index.js

```diff
@@ -0,0 +1,511 @@
+
+		endDate = this.clampDateToRange( endDate, {
+			dateFrom: firstSelectableDate,
+			dateTo: lastSelectableDate,
+		} );
+
+		startDate = isNil( this.props.selectedStartDate )
+			? this.props.moment( endDate ).subtract( 1, 'months' )
+			: this.props.moment( this.props.selectedStartDate );
+
+		startDate = this.clampDateToRange( startDate, {
+			dateFrom: firstSelectableDate,
+			dateTo: lastSelectableDate,
+		} );
+
+		// Ensure start is before end otherwise flip the values
+		if ( endDate.isBefore( startDate ) ) {
+			// flip values via array destructuring (think about it...)
+			[ startDate, endDate ] = [ endDate, startDate ];
+		}
+
+		// Build initial state
+		this.state = {
+			popoverVisible: false,
+			staleStartDate: '',
+			staleEndDate: '',
+			startDate: startDate,
+			endDate: endDate,
+			staleDatesSaved: false,
+			// this need to be independent from startDate because we must independently validate them
+			// before updating the central source of truth (ie: startDate)
+			textInputStartDate: this.formatDateToLocale( startDate ),
+			textInputEndDate: this.formatDateToLocale( endDate ),
+		};
+
+		// Ref to the Trigger <button> used to position the Popover	component
+		this.triggerButtonRef = React.createRef();
+	}
+
+	/**
+	 * Opens the popover
+	 * Note this does not commit the current date state
+	 */
+	openPopover = () => {
+		this.setState( {
+			popoverVisible: true,
+		} );
+	};
+
+	/**
+	 * Closes the popover
+	 * Note this does not commit the current date state
+	 */
+	closePopover = () => {
+		this.setState( {
+			popoverVisible: false,
+		} );
+
+		// As no dates have been explicity accepted ("Apply" not clicked)
+		// we need to revert back to the original cached dates
+		this.revertDates();
+	};
+
+	/**
+	 * Toggles the popover between open/closed
+	 * Note this does not commit the current date state
+	 */
+	togglePopover = () => {
+		if ( this.state.popoverVisible ) {
+			this.closePopover();
+		} else {
+			this.openPopover();
+		}
+	};
+
+	/**
+	 * Updates state with current value of start/end
+	 * text inputs
+	 * @param  {string} val        the value of the input
+	 * @param  {string} startOrEnd either "Start" or "End"
+	 */
+	handleInputChange = ( val, startOrEnd ) => {
+		this.setState( {
+			[ `textInput${ startOrEnd }Date` ]: val,
+		} );
+	};
+
+	/**
+	 * Ensure dates are valid according to standard rules
+	 * and special configuration component config props
+	 *
+	 * @param  {moment}  date MomentJS date object
+	 * @return {Boolean}      whether date is considered valid or not
+	 */
+	isValidDate( date ) {
+		const { firstSelectableDate, lastSelectableDate } = this.props;
+
+		const epoch = this.props.moment( '01/01/1970', this.getLocaleDateFormat() );
+
+		// By default check
+		// 1. Looks like a valid date
+		// 2. after 01/01/1970 (avoids bugs when really stale dates are treated as valid)
+		let validations = date.isValid() && date.isSameOrAfter( epoch );
```

Comment:
`isValid` would better represent a `boolean` value, `validations` would suggest to me either a collection of validation objects (to be used in showing a validation message, for example).

It looks like we could be making use of early returns too to avoid the need for `validations` completely , something like so:
```js
if ( ! date.isValid() || ! date.isSameOrAfter( epoch ) ) {
  return false;
}

if ( firstSelectableDate &&  date.isBefore( firstSelectableDate ) {
  return false;
}

if ( lastSelectableDate &&  date.isAfter( lastSelectableDate ) {
  return false;
}

return true;
```

That allows us to skip some extra checking and reassigning by eagerly rejecting the validation.
As a bonus It also sets us up nicely for the possibility of adding specific validation errors, if the need arises:

```js
if ( ! date.isValid() || ! date.isSameOrAfter( epoch ) ) {
  return new Error( 'some message relating to the above condition' );
}

if ( firstSelectableDate &&  date.isBefore( firstSelectableDate ) {
  return new Error( 'some message relating to the above condition' );
}

if ( lastSelectableDate &&  date.isAfter( lastSelectableDate ) {
  return new Error( 'some message relating to the above condition' );
}

// returning false or just nothing at all would suggest no validation issues here
return false;
```

Reply:
I do think it's easier as a reviewer to spot stuff like this than it is while writing the code itself. It's the same reason I often find myself looking back at my own code with confusion 😄